### PR TITLE
Refine the logic of gitContribute

### PR DESCRIPTION
**Pull request by qinezh**
#951

### CODE REVIEW

File: src/docfx.website.themes/common/ManagedReference.common.js [hunk 1]
```diff
@@ -125,7 +125,7 @@ function handleNamespace(model) {
 function handleItem(vm, gitContribute, gitUrlPattern) {
   // get contribution information
   vm.docurl = common.getImproveTheDocHref(vm, gitContribute, gitUrlPattern);
-  vm.sourceurl = common.getViewSourceHref(vm, null, gitUrlPattern);
+  vm.sourceurl = common.getViewSourceHref(vm, null);
```
Comment (superyyrrzz) [inline src/docfx.website.themes/common/ManagedReference.common.js]: >null [](start = 46, length = 4)

This parameter is gitUrlPattern, why null?

File: src/docfx.website.themes/common/common.js [hunk 2]
```diff
@@ -103,36 +102,64 @@ function normalizeGitUrlToHttps(repo) {
...
+        repoInfo.repo = gitRemote.repo;
+    }
+
+    if (gitContribute && gitContribute.branch) {
+        repoInfo.branch = gitContribute.branch;
+    } else if (gitRemote && gitRemote.branch) {
+        repoInfo.branch = gitRemote.branch;
+    }
+
+    if (gitContribute && gitContribute.path) {
```
Comment (superyyrrzz) [inline src/docfx.website.themes/common/common.js]: >path [](start = 39, length = 4)

seems path has different meaning in "source" and "gitContribute", use another property?

File: src/docfx.website.themes/common/RestApi.common.js [hunk 3]
```diff
@@ -16,7 +16,7 @@ exports.transform = function (model) {
                 child.operation = child.operation.toUpperCase();
             }
             child.path = appendQueryParamsToPath(child.path, child.parameters);
-            child.sourceurl = child.sourceurl || common.getViewSourceHref(child, null, model._gitUrlPattern);
+            child.sourceurl = child.sourceurl || common.getViewSourceHref(child, null);
```
Comment (superyyrrzz) [inline src/docfx.website.themes/common/RestApi.common.js]: >getViewSourceHref [](start = 56, length = 17)

also update

File: src/docfx.website.themes/common/common.js [hunk 4]
```diff
@@ -103,36 +102,70 @@ function normalizeGitUrlToHttps(repo) {
...
+        modelGitInfo.repo = gitRemote.repo;
+    }
+
+    if (gitContribute && gitContribute.branch) {
+        modelGitInfo.branch = gitContribute.branch;
+    } else if (gitRemote && gitRemote.branch) {
+        modelGitInfo.branch = gitRemote.branch;
+    }
+
+    if (gitContribute) {
```
Comment (qinezh) [inline src/docfx.website.themes/common/common.js]: nice catch

---
In reply to: [91652863](https://github.com/dotnet/docfx/pull/979#discussion_r91652863) [](ancestors = 91652863)

File: src/docfx.website.themes/common/common.js [hunk 5]
```diff
@@ -155,9 +158,7 @@ function getGitContributeFallbackWithGitRemote(gitRemote, gitContribute) {
     }
 
     if (gitContribute) {
-        // apiSpecFolder defines the folder contains overwrite files for MRef
-        // the default value is apiSpec
-        modelGitInfo.apiSpecFolder = gitContribute.path || "apiSpec";
+        modelGitInfo.apiSpecFolder = gitContribute.path;
```
Comment (superyyrrzz) [inline src/docfx.website.themes/common/common.js]: What about gitContribute is defined, while gitContribute.path is undefined

Comment (qinezh) [inline src/docfx.website.themes/common/common.js]: different implementation same effect?

---
In reply to: [91675728](https://github.com/dotnet/docfx/pull/979#discussion_r91675728) [](ancestors = 91675728)

**Comment by superyyrrzz** [comment]
e2e failed: http://docfx-ci-0.cloudapp.net/viewLog.html?buildId=6542&buildTypeId=DocfxCiWithScripts_DocfxCiForDevBranch&tab=buildLog#_focus=3509

File: src/docfx.website.themes/common/common.js [hunk 6]
```diff
@@ -28,16 +28,15 @@ function getHtmlId(input) {
     return input.replace(/\W/g, '_');
 }
 
-function getViewSourceHref(item, gitContribute, gitUrlPattern) {
```
Comment (vicancy) [inline src/docfx.website.themes/common/common.js]: or rather, keep the interface unchanged...

---
In reply to: [91878401](https://github.com/dotnet/docfx/pull/979#discussion_r91878401) [](ancestors = 91878401)

**Comment by superyyrrzz** [comment]
:shipit:

File: src/docfx.website.themes/common/common.js [hunk 7]
```diff
@@ -103,36 +103,71 @@ function normalizeGitUrlToHttps(repo) {
...
+    if (modelGitInfo.repo.substr(-4) === '.git') {
+        modelGitInfo.repo = modelGitInfo.repo.substr(0, modelGitInfo.repo.length - 4);
     }
 
-    var patternName = getPatternName(repo, gitUrlPattern);
+    var patternName = getPatternName(modelGitInfo.repo, gitUrlPattern);
     if (!patternName) return '';
-    return gitUrlPatternItems[patternName].generateUrl(repo, branch, remote.path, startLine);
+
+    modelGitInfo.startLine = startLine;
```
Comment (vicancy) [inline src/docfx.website.themes/common/common.js]: >modelGitInfo [](start = 4, length = 12)

naming: gitInfo or gitInfoModel?

File: src/docfx.website.themes/common/common.js [hunk 8]
```diff
@@ -103,36 +103,71 @@ function normalizeGitUrlToHttps(repo) {
...
+    return gitUrlPatternItems[patternName].generateUrl(modelGitInfo);
+}
+
+function getGitContributeFallbackWithGitRemote(gitRemote, gitContribute) {
+    var modelGitInfo = {
+        // apiSpecFolder defines the folder contains overwrite files for MRef, the default value is apiSpec
+        apiSpecFolder: "apiSpec"
+    };
+
+    if (gitContribute && gitContribute.repo) {
```
Comment (vicancy) [inline src/docfx.website.themes/common/common.js]: >if [](start = 4, length = 2)

is it clearer to:
if (!gitContribute){
...
}
if (!gitRemote){
...
}

return {
   repo: gitContribute.repo || gitRemote.repo
   ...
}

File: src/docfx.website.themes/common/common.js [hunk 9]
```diff
@@ -103,36 +103,71 @@ function normalizeGitUrlToHttps(repo) {
...
-
-    if (repo == undefined || branch == undefined) return '';
+    var modelGitInfo = getGitContributeFallbackWithGitRemote(remote, gitContribute);
+    if (!modelGitInfo.repo || !modelGitInfo.branch || !modelGitInfo.path) {
+        return '';
+    }
 
-    if (repo.substr(-4) === '.git') {
-        repo = repo.substr(0, repo.length - 4);
+    if (modelGitInfo.repo.substr(-4) === '.git') {
```
Comment (vicancy) [inline src/docfx.website.themes/common/common.js]: >modelGitInfo.repo. [](start = 8, length = 18)

could be null

File: src/docfx.website.themes/common/common.js [hunk 10]
```diff
@@ -103,36 +103,71 @@ function normalizeGitUrlToHttps(repo) {
...
-    var patternName = getPatternName(repo, gitUrlPattern);
+    var patternName = getPatternName(modelGitInfo.repo, gitUrlPattern);
     if (!patternName) return '';
-    return gitUrlPatternItems[patternName].generateUrl(repo, branch, remote.path, startLine);
+
+    modelGitInfo.startLine = startLine;
+    return gitUrlPatternItems[patternName].generateUrl(modelGitInfo);
+}
+
+function getGitContributeFallbackWithGitRemote(gitRemote, gitContribute) {
```
Comment (qinezh) [inline src/docfx.website.themes/common/common.js]: agree

---
In reply to: [92736456](https://github.com/dotnet/docfx/pull/979#discussion_r92736456) [](ancestors = 92736456)

File: src/docfx.website.themes/common/common.js [hunk 11]
```diff
@@ -64,34 +64,34 @@ function isRelativePath(path) {
 var gitUrlPatternItems = {
     'github': {
         'testRegex': /^(https?:\/\/)?(\S+\@)?(\S+\.)?github\.com(\/|:).*/i,
-        'generateUrl': function(repo, branch, path, startLine) {
-            var url = normalizeGitUrlToHttps(repo);
-            url += '/blob' + '/' + branch + '/' + path;
-            if (startLine && startLine > 0) {
-                url += '/#L' + startLine;
+        'generateUrl': function (modelGitInfo) {
+            var url = normalizeGitUrlToHttps(modelGitInfo.repo);
```
Comment (qinezh) [inline src/docfx.website.themes/common/common.js]: yep, the caller could guarantee it.

---
In reply to: [92736776](https://github.com/dotnet/docfx/pull/979#discussion_r92736776) [](ancestors = 92736776)

**Comment by ansyral** [comment]
:shipit:

File: src/docfx.website.themes/common/common.js [hunk 12]
```diff
@@ -103,36 +103,69 @@ function normalizeGitUrlToHttps(repo) {
...
-    var patternName = getPatternName(repo, gitUrlPattern);
+    var patternName = getPatternName(gitInfo.repo, gitUrlPattern);
     if (!patternName) return '';
-    return gitUrlPatternItems[patternName].generateUrl(repo, branch, remote.path, startLine);
+
+    gitInfo.startLine = startLine;
+    return gitUrlPatternItems[patternName].generateUrl(gitInfo);
+}
+
+function getGitContributeFallbackWithGitRemote(gitContribute, gitRemote) {
```
Comment (vicancy) [inline src/docfx.website.themes/common/common.js]: simply getGitInfo?

**Pull request merged**